linker.go

```diff
@@ -277,3 +278,60 @@ func fixupProbeReadKernel(ins *asm.Instruction) {
 		ins.Constant = int64(asm.FnProbeReadStr)
 	}
 }
+
+// resolveKconfigReferences creates and populates a .kconfig map if necessary.
+//
+// Returns a nil Map and no error if no references exist.
+func resolveKconfigReferences(insns asm.Instructions) (*Map, error) {
+	var spec *MapSpec
+	iter := insns.Iterate()
+	for iter.Next() {
+		ins := iter.Ins
+
+		spec, _ = ins.Metadata.Get(kconfigMeta{}).(*MapSpec)
+		if spec == nil {
+			continue
+		}
+
+		goto populate
+	}
+
+	return nil, nil
+
+populate:
+	cpy := spec.Copy()
+	if err := resolveKconfig(cpy); err != nil {
+		return nil, err
+	}
+
+	kconfig, err := NewMap(cpy)
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		insSpec, _ := iter.Ins.Metadata.Get(kconfigMeta{}).(*MapSpec)
+		if insSpec == nil {
+			goto next
+		}
+
+		if insSpec != spec {
+			kconfig.Close()
+			return nil, fmt.Errorf("instruction %d: reference to multiple .kconfig maps is not allowed", iter.Index)
+		}
+
+		if err := iter.Ins.AssociateMap(kconfig); err != nil {
```

Comment:
Only half joking: we could implement this without going via a map in the first place and just patch instructions directly. Something for the future maybe.

Reply:
Instead of putting `LINUX_VERSION_CODE` into the map at an offset, and referencing the map from this instruction, we could rewrite the instruction to be a `LdImm64` of LINUX_VERSION_CODE. This avoids the indirection via the map and works on older kernels. It has subtleties though when a single value is accessed multiple times, etc.

Reply:
I'm not going to do it now, since I don't have a use case yet. It might happen that we need something like it in cilium though. Then we can probably extend this to anything that fits in 64 bit.

Reply:
> It has subtleties though when a single value is accessed multiple times, etc.

Indeed, but it's also a subtlety that cannot realistically be worked around by the library. It would require userspace to track all accesses to the register the map pointer was loaded into, accounting for branches and jumps, as well as the possibility of aliasing/moving, etc. This is best left to the verifier. :slightly_smiling_face: